### DISCUSSION TITLE
How do I access the `dist/package_x/managed` folder ?

### DISCUSSION THREAD

**Discussion by jeromelaban** [unoplatform/uno#9001] [Q&A]
_[Question from discord](https://discord.com/channels/372137812037730304/550416151172087808/984516719294959666)_

Heyo! For some reason, accessing the `dist/package_x/managed` folder in WASM doesn't work. I'm trying to access the `Appname.Wasm.clr` file with the same way how it was done for the sample project that showcased compiling C# in wasm and accessing the `dotnet-sdk` folders pre-bundled `.clr`  files. Other folders work fine though, as they are in the `.csproj` as content. 

Is accessing the managed folder blocked by Uno or is there some other way to get it?

**Comment by jeromelaban** [comment]
Why do you need to get the file there? Is it because you're building a compiler in your app?

**Reply by Cyroxin** [reply]
Heya! I'm not building a compiler, but bundling a compiler in a similar way the sample project has done. 

I have an interface within the source of the application:
``` C#
public interface IVisual
	{
		public string Name { get; set; }
		public Uri Cover { get; set; }
	}
```

I am also downloading and compiling a source file from the web that implements the interface:
``` C#
public class VisualDownloadedFromTheWeb : IVisual
	{
		public string Id { get; set; }
		public string Name { get; set; }
		public Uri Cover { get; set; }
	}
```

To make this compile, I add the following as a reference to [Compilation.AddReferences](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.compilation.addreferences):

``` C#
typeof(IVisual).Assembly.Location
```

However, this only works on non-WASM targets. Since on webassembly `` Assembly.Location`` defaults to an empty string.

____

Why is it important to be able to add it as a reference instead of a SyntaxTree you may ask? Well, because of this:

``` C#
        // Find all implementations of IVisual within the loaded assembly
	Type[] iLoadTypes = (from t in assembly.GetExportedTypes()
						where !t.IsInterface && !t.IsAbstract
						where typeof(IVisual).IsAssignableFrom(t)
						select t).ToArray();
        // Create an instance of the class that implements IVisual
	return (IVisual) Activator.CreateInstance(iLoadTypes.FirstOrDefault());

```
If the compilation were to rely on an SyntaxTree, ```typeof(IVisual).IsAssignableFrom(t)```would return false since there are now effectively two IVisual implementations. The InMemoryAssembly one and the one the application has. Although both of the IVisual implementations are identical, the above code will not work if compilation relies on a SyntaxTree.

Using references to actual code within the application will allow for a simple use such as this:
``` C#
var Name = Instance.Name;
```

**Reply by jeromelaban** [reply]
I see. The files located in the `managed` folder are accessible, but it's generally not a good idea to go read those files manually through their url because it's not cross platform compatible. Now, it may be the only way for you to get what you need because as you've found, the "Location" of the assembly is null to avoid consuming twice the memory when the runtime initializes (one for the VFS, and one for the loaded assembly in the runtime).

As a workaround, you can use the [`UNO_BOOTSTRAP_APP_BASE`](https://platform.uno/docs/articles/external/uno.wasm.bootstrap/doc/features-environment-variables.html#configuration-environment-variables) environment variable to get it. But it may break in a future bootstrapper version.

**Reply by Cyroxin** [reply]
Heya!

I believe I am missing something. I've attempted the following among various others:

``` C#
var AppPackage = Environment.GetEnvironmentVariable("UNO_BOOTSTRAP_APP_BASE");
var targetFile = Path.Combine(AppPackage, "managed","Appname.Wasm.clr");

var AppAssembly = await StorageFile.GetFileFromPathAsync(targetFile);

using var stream = await AppAssembly.OpenReadAsync();
var streamCopy = new MemoryStream();
stream.AsStream().CopyTo(streamCopy);

streamCopy.Position = 0;
return MetadataReference.CreateFromStream(streamCopy);
```
``` C#
var AppPackage = Environment.GetEnvironmentVariable("UNO_BOOTSTRAP_APP_BASE");
var targetFile = Path.Combine(AppPackage, "managed","Appname.Wasm.clr");

var AppAssembly = await StorageFile.GetFileFromApplicationUriAsync(new Uri($"ms-appx:///{targetFile}"));

using var stream = await AppAssembly.OpenReadAsync();
var streamCopy = new MemoryStream();
stream.AsStream().CopyTo(streamCopy);

streamCopy.Position = 0;
return MetadataReference.CreateFromStream(streamCopy);
```

**Reply by jeromelaban** [reply]
That's what I meant by the "not cross platform" part, you'll need to use an `HttpClient` to get to these files.

**Reply by Cyroxin** [reply]
Heya!

Thank you! Here is the code for others too:

``` C#
var AppPackage = Environment.GetEnvironmentVariable("UNO_BOOTSTRAP_APP_BASE");
var targetFile = Path.Combine("https://localhost:54411", AppPackage, "managed", "Appname.Wasm.clr");

HttpClient client = new HttpClient();
var stream = await client.GetStreamAsync(targetFile);
return MetadataReference.CreateFromStream(stream);
```

**Reply by Cyroxin** [reply]
As a note to future readers,

I've been looking at better solutions and found a much better method than accessing the managed folder, which may change overtime in the way it is accessed. 

For the application to be able to access the same assembly as the compiled **InMemoryAssembly**, I found that it is best to create a separate **Class Library** where to store all interfaces or other commonly shared classes. The Uno shared project does not create a .dll file, so they cannot be placed there. 

It is then as simple as adding a dependency on the **Class Library** to your application and then running the application that is referencing the **Class Library**. An assembly called ``NameOfYourClassLibrary.dll`` is created, which needs to be placed inside the folder of assemblies included in compiling the InMemoryAssembly, without forgetting to add the new assembly path to the list of reference paths to include in compilation.
